zcash_client_backend/src/decrypt.rs

```diff
@@ -30,7 +31,8 @@ pub struct DecryptedOutput {
 
 /// Scans a [`Transaction`] for any information that can be decrypted by the set of
 /// [`ExtendedFullViewingKey`]s.
-pub fn decrypt_transaction(
+pub fn decrypt_transaction<P: consensus::Parameters>(
+    parameters: &P,
```

Comment:
To my previous point: If we were using a closed-world `enum Parameters { Main, Test }` then we would definitely need an explicit `parameters` value passed into this (and every other) API. With the open-world `trait Parameters` design that we currently use (which allows for arbitrary networks to be defined), without `&self` in any of its methods (which as above is not AFAICT a required change), the `<P: consensus::Parameters>` generic parameter alone will suffice, and would reduce the scope of changes in this PR.

tl;dr If `&self` is not required inside `Parameters`, then remove the `parameters` value from this and all other APIs where it has been introduced (leaving the new `<P: consensus::Parameters>` bounds), and change e.g. `parameters.is_nu_active(..)` to `P::is_nu_active(..)`.

Reply:
This is implemented as an open-world constraint; so long as P implements `Parameters` it doesn't matter whether that's a MainNet `P` or a TestNet `P` - I think that trying to smuggle mainnet vs. testnet to a compile-time switch is anti-modular; explicit passing of a Parameters value is the correct thing to do here.

Reply:
I think you are mixing several separate issues together:

- `zcash_client_sqlite` (where this particular function lives) uses a `mainnet` feature flag to switch at compile time between mainnet and testnet support. This is a decision that made sense at the time, but may be worth revisiting to move to runtime configuration. However, doing so is completely orthogonal to this PR.

- The presence or absence of `parameters: &P` makes zero difference to the way in which this API can be used. In either case, `decrypt_transaction` is monomorphized based on `P`, so passing in a `parameter` value necessarily bounds the function to a type at the caller:
  ```
  decrypt_transaction(&MainNet, ...) === decrypt_transaction::<MainNet>(&MainNet, ...)
  ```
  This also means that adding the superfluous `parameter` value makes the API as-is more confusing, as it can't actually be used like a parameter at runtime. The following will [(correctly) fail to compile](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fe3c11bcda5018330950f73a2f5b7d56), because `MainNet` and `TestNet` are different types:
  ```rust
  let parameters = if config.is_mainnet() {
      MainNet
  } else {
      TestNet
  };
  decrypt_transaction(&parameters, ...)
  ```
  [Boxing would be necessary to use this pattern](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a05e6f830ada99ef02a3edfb68f67f4b), along with providing a blanket `impl Parameters for Box<dyn Parameters>`.

  As-is, the runtime usage for this particular function [would be](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=abc3f2bf1e74620c1ad8520a87530cab):
  ```rust
  if config.is_mainnet() {
      decrypt_transaction::<MainNet>(...)
  } else {
      decrypt_transaction::<TestNet>(...)
  };
  ```
- What I think you're actually trying to suggest be done here is `decrypt_transaction(parameters: &dyn Parameters)`, using dynamic dispatch for all use cases instead of static dispatch. This may indeed be a useful improvement to the API, but it's also completely orthogonal to this PR, and should be explored separately.

Reply:
A way to avoid the boxing would be to implement `Parameters` for an underlying sum type such as you suggested; the callee still gets to remain ignorant of the type which it is provided.

I think what I'm uncomfortable with is the basic idea of using the type system to "smuggle" a constant value via monomorphization. I don't dispute that it works, but I think that it's morally equivalent to using globals for the same purpose - compile-time globals, to be sure, and not mutable ones, so some of the downsides are removed - but I still think the design is anti-modular because any caller may "fix" the type (as we see below with consensus::MainNet) and then you *have* to catch it in review because the type system can no longer help another, outer caller determine that their callee will only function properly on a particular network. 

More importantly, it precludes a user of the library from implementing, for example, a flag to be read on startup which determines which network the resulting binary will run on - and means that fundamentally, we're not able to run the same binary on testnet as on mainnet, which kind of defeats the purpose of reproducible builds. What does it matter if you can reproducibly create a build for mainnet, if you can never have run that same binary on testnet?

Reply:
I agree that it may be superfluous with respect to this PR, but I'm going to submit a PR to replace these static determinations with the well established approach of passing values to functions. :)

Reply:
You're both right:
* The Parameters argument is of no use in this PR.
* The choice of network should be dynamic.

Changing it to dynamic in a separate PR is the right approach. If the parameters argument is removed in the meantime, it will actually make that change more regular.

Reply:
Another reason to do the dynamic change sir stately is that we're under time pressure to get this ZIP 212 change merged so that it can be audited.

Reply:
PR submitted: https://github.com/zcash/librustzcash/pull/261